File: app/src/main/java/art/manguste/android/gamesearch/GamesListFragment.java
package art.manguste.android.gamesearch;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ProgressBar;

import java.util.ArrayList;
import java.util.List;

import art.manguste.android.gamesearch.core.Game;
import art.manguste.android.gamesearch.db.GameDatabase;
import art.manguste.android.gamesearch.threads.GamesApiLoader;
import art.manguste.android.gamesearch.core.SearchType;
import art.manguste.android.gamesearch.db.FavoriteGame;

import static art.manguste.android.gamesearch.api.URLMaker.formURL;
import static art.manguste.android.gamesearch.api.JsonParser.parseGameData;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link GamesListFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class GamesListFragment extends Fragment {

    private static final String TAG = GamesListFragment.class.getSimpleName()+"CheckLoader";

    private static final String SEARCH_TYPE = "search_type";
    private static final int LOADER_BY_NAME_ID = 1;
    private static final int LOADER_HOT_ID = 2;

    private ProgressBar mProgressBar;
    private EditText mSearchByNameTextView;
    private GameCardAdapter mAdapter;
    private RecyclerView mRecyclerView;
    private SearchType searchType;
    LoaderManager mLoaderManager;
    private GameDatabase gameDatabase;

    public GamesListFragment() {
        // Required empty public constructor
    }

    /**
     * New instance via factory method
     * this fragment using the provided parameters.
     *
     * @param searchType - determines what kind of search and API request to use
     * @return new instance of fragment GamesListFragment.
     */
    public static GamesListFragment newInstance(SearchType searchType) {
        GamesListFragment fragment = new GamesListFragment();
        Bundle args = new Bundle();
        args.putString(SEARCH_TYPE, String.valueOf(searchType));
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            searchType = SearchType.valueOf(getArguments().getString(SEARCH_TYPE));
        }
        mLoaderManager = LoaderManager.getInstance(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, final ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_game_search, container, false);
        mProgressBar = view.findViewById(R.id.pb_loading);
        mSearchByNameTextView = view.findViewById(R.id.et_search_by_name);

        // Recycler view stuff
        mRecyclerView = view.findViewById(R.id.rv_games_list);
        int imageSize = getResources().getDimensionPixelSize(R.dimen.icon_size) * 2;

        mAdapter = new GameCardAdapter(getContext(), imageSize, container, searchType);
        mRecyclerView.setAdapter(mAdapter);

        // connect data and view
        mRecyclerView.setLayoutManager(new GridLayoutManager(getActivity(), 1));

        if (!SearchType.SEARCH.equals(searchType)){
            (view.findViewById(R.id.ll_search_by_name)).setVisibility(View.GONE);
        }

        // on a click "start search" button
        (view.findViewById(R.id.btn_start_search)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startGameSearch(true);
            }
        });

        Log.d(TAG, "onCreateView " + searchType.toString());

        //mLoaderManager.initLoader(LOADER_HOT_ID, null, this);
        gameDatabase = GameDatabase.getInstance(getContext());
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.d(TAG, "onViewCreated " + searchType.toString());
        // TODO try to change on ViewModel
        if (!SearchType.SEARCH.equals(searchType)){
            startGameSearch(false);
        }
    }


    /**
     * Go to API and grab data
     * */
    private void startGameSearch(boolean forceNewRequest) {
        if (SearchType.HOT.equals(searchType)) {
            if (mLoaderManager.getLoader(LOADER_HOT_ID) == null){
                mLoaderManager.initLoader(LOADER_HOT_ID, null, mAPILoaderCallbacks);
            }
        } else if (SearchType.SEARCH.equals(searchType)){
            if (mLoaderManager.getLoader(LOADER_BY_NAME_ID) == null){
                mLoaderManager.initLoader(LOADER_BY_NAME_ID, null, mAPILoaderCallbacks);
            } else if (forceNewRequest){
                mLoaderManager.restartLoader(LOADER_BY_NAME_ID, null, mAPILoaderCallbacks);
            }
        } else if (SearchType.FAVORITE.equals(searchType)){
            getFavoriteGames();
        }
    }

    private final LoaderManager.LoaderCallbacks<ArrayList<Game>> mAPILoaderCallbacks =
            new LoaderManager.LoaderCallbacks<ArrayList<Game>>() {

                @Override
                @NonNull
                public Loader<ArrayList<Game>> onCreateLoader(int id, @Nullable Bundle args) {
                    mProgressBar.setVisibility(View.VISIBLE);
                    Log.d(TAG, "onCreateLoader " + searchType.toString() +" with Context="+getContext());

                    String searchTxt = String.valueOf(mSearchByNameTextView.getText());
                    String urlString = formURL(searchType, searchTxt);
                    return new GamesApiLoader(getContext(), urlString, searchType);
                }

                @Override
                public void onLoadFinished(@NonNull Loader<ArrayList<Game>> loader, ArrayList<Game> data) {
                    mProgressBar.setVisibility(View.GONE);

                    //TODO if none games found - set text about it

                    // change data in view
                    if (data != null && !data.isEmpty()) {
                        mAdapter.setGames(data);
                    }
                    Log.d(TAG, "onLoadFinished " + searchType.toString());
                }

                @Override
                public void onLoaderReset(@NonNull Loader<ArrayList<Game>> loader) {
                    //Log.d(TAG, "onLoaderReset " + searchType.toString());
                    //mAdapter.setGames(null);
                }
            };

    // for Favorite Games Fragment
    // LiveData + observe
    private void getFavoriteGames() {
        LiveData<List<FavoriteGame>> favGames
                = (LiveData<List<FavoriteGame>>) gameDatabase.favoriteGameDao().selectAll();

        favGames.observe(getActivity(), new Observer<List<FavoriteGame>>() {
            @Override
            public void onChanged(List<FavoriteGame> favoriteGames) {
                ArrayList<Game> games = new ArrayList();
                for (FavoriteGame favGame : favoriteGames) {
                    if (favGame.getJson() != null){
                        games.add(parseGameData(favGame.getJson(), true));
                    }
                }

                // change data in view
                if (games != null && !games.isEmpty()) {
                    mAdapter.setGames(games);
                }

            }

        });
    }
}

File: app/src/main/java/art/manguste/android/gamesearch/gamesList/GamesListFragment.kt
package art.manguste.android.gamesearch.gamesList

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import art.manguste.android.gamesearch.GameDetailActivity
import art.manguste.android.gamesearch.core.GameBriefly
import art.manguste.android.gamesearch.core.LoadStatus
import art.manguste.android.gamesearch.core.SearchType
import art.manguste.android.gamesearch.databinding.FragmentGameSearchBinding
import art.manguste.android.gamesearch.db.GameDatabase

import com.google.android.material.snackbar.Snackbar

/**
 * A simple [Fragment] subclass.
 * Use the [GamesListFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class GamesListFragment : Fragment() {

/*    private val gamesListVM: GamesListVM by lazy {
        ViewModelProvider(this).get(GamesListVM::class.java)
    }*/
    private lateinit var gamesListVM: GamesListVM

    private var searchType: SearchType? = null
    private lateinit var binding: FragmentGameSearchBinding

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        binding = FragmentGameSearchBinding.inflate(inflater)

        val application = requireNotNull(this.activity).application
        val dataSource = GameDatabase.getInstance(application).gameDao
        gamesListVM = GamesListVM(application, dataSource)

        // data binding will observe LiveData with the lifecycle of the fragment
        binding.lifecycleOwner = this
        // access to View Model from the layout

        binding.recyclerGames.adapter = GameAdapter(OnClickListener(
                // on click by whole card - going to the detail info about game
                { game ->
                    toDetailInfo(game)
                },
                // on click "make favorite" button - make favorite
                { game ->
                    changeGamesFavoriteStatus(game)
                }))

        setHasOptionsMenu(true)
        return binding.root
    }

    private fun toDetailInfo(game: GameBriefly){
        Log.d(TAG, "OnClick ${game.name}")
        val intent = Intent(activity, GameDetailActivity::class.java)
                .putExtra(GameDetailActivity.EXTRA_GAME_CODE, game.alias)
                .putExtra(GameDetailActivity.EXTRA_GAME_NAME, game.name)

        Log.d(TAG, "OnClick start activity ${game.name}")
        startActivity(intent)
    }

    private fun changeGamesFavoriteStatus(game: GameBriefly)  {
        Log.d(TAG, "OnClickFavorite ${game.name}")

        // form message into snakbar
        val snackMessage = if (game.isFavorite) "${game.name} removed from favourites"
        else "${game.name} added to favourites"

        // db actions
        val application = requireNotNull(this.activity).application
        val dataSource = GameDatabase.getInstance(application).gameDao
        // form actions
        when (game.isFavorite) {
            true -> {
                game.isFavorite = false
                gamesListVM.removeGameFavorite(game, dataSource)
            }
            false -> {
                game.isFavorite = true
                gamesListVM.addGameFavorite(game, dataSource)
            }
        }

        Snackbar.make(requireView(), snackMessage, Snackbar.LENGTH_LONG).show()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)



        gamesListVM.status.observe(viewLifecycleOwner, { status ->
            when (status!!) {
                LoadStatus.LOADING -> {
                    binding.progressBar.visibility = View.VISIBLE
                    binding.netLostImage.visibility = View.GONE
                }
                LoadStatus.DONE -> {
                    binding.progressBar.visibility = View.GONE
                    binding.netLostImage.visibility = View.GONE
                }
                LoadStatus.ERROR -> {
                    binding.progressBar.visibility = View.GONE
                    binding.netLostImage.visibility = View.VISIBLE
                }
                LoadStatus.NONE -> {
                    binding.progressBar.visibility = View.GONE
                    binding.netLostImage.visibility = View.GONE
                }
            }
        })

        searchType = requireArguments().getString(SEARCH_TYPE)?.let { SearchType.valueOf(it) }
        Log.d(TAG, "SEARCH_TYPE = $searchType")

        // hide search panel if it needed
        when (searchType)  {
            SearchType.HOT -> {
                binding.panelSearchGame.visibility = View.GONE
                gamesListVM.getHotGamesList()
            }
            SearchType.SEARCH -> {
                binding.panelSearchGame.visibility = View.VISIBLE
                // search Game by click on button
                binding.btnStartSearch.setOnClickListener {
                    binding.searchByTitle.text.isNotEmpty().let {
                        gamesListVM.getSearchGameList(binding.searchByTitle.text.toString())
                    }
                }
            }
            SearchType.FAVORITE -> {
                binding.panelSearchGame.visibility = View.GONE
                val application = requireNotNull(this.activity).application  // todo - del this madness, its just for test idea
                val dataSource = GameDatabase.getInstance(application).gameDao
                gamesListVM.getDBGameList(dataSource)
            }
            else -> Log.d(TAG, "Unexpected search type = $searchType")
        }

        //if (searchType != SearchType.FAVORITE){
            gamesListVM.gamesList.observe(viewLifecycleOwner, { games ->
                Log.d(TAG, "games = ${games.size}")
                (binding.recyclerGames.adapter as GameAdapter).apply {
                    reloadGames(games)
                }
            })
        //}
    }

    companion object {
        private val TAG = GamesListFragment::class.java.simpleName
        private const val SEARCH_TYPE = "search_type"

        /**
         * New instance via factory method
         * this fragment using the provided parameters.
         *
         * @param searchType - determines what kind of search and API request to use
         * @return new instance of fragment GamesListFragment.
         */
        fun newInstance(searchType: SearchType): GamesListFragment {
            val fragment = GamesListFragment()
            val args = Bundle()
            args.putString(SEARCH_TYPE, searchType.toString())
            fragment.arguments = args
            return fragment
        }
    }
}